src/transport/mqtt/connection.rs

```diff
@@ -0,0 +1,892 @@
+    store::{error::StoreError, wrapper::StoreWrapper, PropertyStore, StoredProp},
+    transport::mqtt::{pairing::ApiClient, payload::Payload, AsyncClientExt},
+};
+
+use super::{
+    client::{AsyncClient, EventLoop},
+    config::transport::TransportProvider,
+    ClientId, PayloadError, SessionData,
+};
+
+/// Errors while initializing the MQTT connection.
+#[non_exhaustive]
+#[derive(thiserror::Error, Debug)]
+enum InitError {
+    /// Couldn't send the message.
+    #[error("couldn't send messages for {ctx}")]
+    Client {
+        #[source]
+        backtrace: ClientError,
+        ctx: &'static str,
+    },
+    /// Couldn't wait for the message acknowledgment.
+    #[error("couldn't wait the message acknowledgment for {ctx}")]
+    Notice {
+        #[source]
+        backtrace: NoticeError,
+        ctx: &'static str,
+    },
+    /// Couldn't serialize the device property payload.
+    #[error("coudln't serialize the device property payload")]
+    Payload(#[from] PayloadError),
+}
+
+impl InitError {
+    const fn client(ctx: &'static str) -> impl Fn(ClientError) -> InitError {
+        move |backtrace: ClientError| InitError::Client { backtrace, ctx }
+    }
+
+    const fn notice(ctx: &'static str) -> impl Fn(NoticeError) -> InitError {
+        move |backtrace: NoticeError| InitError::Notice { backtrace, ctx }
+    }
+}
+
+/// MQTT connection to Astarte that can be pulled to receive packets.
+#[derive(Debug)]
+pub(crate) struct MqttConnection {
+    connection: Connection,
+    /// Queue for the packet published while during the connection.
+    buff: VecDeque<Publish>,
+    state: State,
+}
+
+impl MqttConnection {
+    pub(crate) fn new(
+        client: AsyncClient,
+        eventloop: EventLoop,
+        provider: TransportProvider,
+        state: impl Into<State>,
+    ) -> Self {
+        let connection = Connection {
+            client,
+            eventloop: SyncWrapper::new(eventloop),
+            provider,
+        };
+
+        Self {
+            connection,
+            buff: VecDeque::new(),
+            state: state.into(),
+        }
+    }
+
+    /// Iterate for the next publish event from the connection.
+    ///
+    /// Return [`None`] when disconnected.
+    pub(crate) async fn next_publish(&mut self) -> Option<Publish> {
+        // Check if there are backed up publishes.
+        if let Some(publish) = self.buff.pop_front() {
+            return Some(publish);
+        }
+
+        // Here we only get the connected state so we don't need to pass all the arguments, like the
+        // interfaces or the store.
+        let State::Connected(connected) = &mut self.state else {
+            return None;
+        };
+
+        loop {
+            match connected.poll(&mut self.connection).await {
+                Next::Same => {}
+                Next::Publish(publish) => return Some(publish),
+                Next::State(next) => {
+                    debug_assert!(!next.is_connected());
```

Comment:
is it used for a test/mock expectation?

Reply:
No, it's just an assert. 

If you read the line above, there is a let else for the state that must be connected. Here, we make sure that we never return to a next state that is the same as the current, since we should have returned a Next::Same. 

Furthermore, if you look below, this match branch would return from the loop, but we must return only when disconnected or in other words a next state different from connected.

Reply:
Thank you for the explanation, now it's clearer :)